common/src/main/kotlin/streams/service/errors/ErrorService.kt

```diff
@@ -0,0 +1,118 @@
+                is ByteArray -> v
+                else -> v.toString().toByteArray(Charsets.UTF_8)
+            }
+        } catch (e:Exception) {
+            null
+        }
+    }
+    fun toLogString() =
+            """
+ErrorData(originalTopic=$originalTopic, timestamp=$timestamp, partition=$partition, offset=$offset, exception=$exception, key=${key?.toString(Charsets.UTF_8)}, value=${value?.sliceArray(0..Math.min(value.size,200)-1)?.toString(Charsets.UTF_8)}, executingClass=$executingClass)
+    """.trimIndent()
+
+}
+
+abstract class ErrorService(private val config: Map<String, Any> = emptyMap()) {
+
+    data class ErrorConfig(val fail:Boolean=false, val log:Boolean=false, val logMessages:Boolean=false,
+                           val dlqTopic:String? = null, val dlqHeaderPrefix:String = "", val dlqHeaders:Boolean = false, val dlqReplication: Int? = 3) {
+
+        /*
+        https://www.confluent.io/blog/kafka-connect-deep-dive-error-handling-dead-letter-queues
+            "errors.retry.timeout": "-1",
+            "errors.retry.delay.max.ms": "1000",
+
+            "errors.tolerance": "all", "none" == fail-fast, abort sink task
+
+            fail-fast for configuration errors (e.g. validate cypher statements on start)
+            errors.tolerance = all -> silently ignore all bad messages
+
+            org.apache.kafka.connect.runtime.errors.RetryWithToleranceOperator.execAndHandleError(RetryWithToleranceOperator.java
+
+
+            "errors.log.enable": true,
+            "errors.deadletterqueue.context.header.enable"=true/false
+            "errors.deadletterqueue.topic.name": "test-error-topic",
+            "errors.deadletterqueue.topic.replication.factor": 1,
+            "errors.log.include.messages": true,
+        */
+
+        companion object {
+            const val TOLERANCE = "errors.tolerance"
+            const val LOG = "errors.log.enable"
+            const val LOG_MESSAGES = "errors.log.include.messages"
+            const val DLQ_TOPIC = "errors.deadletterqueue.topic.name"
+            const val DLQ_HEADERS = "errors.deadletterqueue.context.header.enable"
+            const val DLQ_HEADER_PREFIX = "errors.deadletterqueue.context.header.prefix"
+            const val DLQ_REPLICATION = "errors.deadletterqueue.topic.replication.factor"
+
+            fun from(props: Properties) = from(props.toMap() as Map<String, Any>)
+
+            fun boolean(v:Any?) = when (v) {
+                null -> false
+                "true" -> true
+                "false" -> false
+                is Boolean -> v
+                else -> false
+            }
+            fun int(v:Any?) = when (v) {
+                null -> 0
+                is Int -> v
+                is String -> v.toInt()
+                else -> 0
+            }
+
+            fun from(config: Map<String, Any?>) =
+                    ErrorConfig(
+                            fail = config.getOrDefault(TOLERANCE, "none") == "none",
```

Comment:
It's really, really easy for beginning users to screw up config when they're going fast and don't fully understand things.  This line has default behavior that fail is false unless tolerance = "none".  The actual valid values for tolerance are "all" or "none".  It'd be really nice to have a startup log message if the config value doesn't match the valid options -- this would really help debug things when users come in with problems, if the plugin tells you that it knows it's badly configured or configured in some way that has an effect the user doesn't expect.